document/core/exec/runtime.rst

```diff
@@ -702,38 +696,115 @@ Its form and meaning is outside the scope of this specification.
+It consists of the current :ref:`shared store <syntax-sharedstore>` and a set of executing *threads*.
+
+A thread consists of a local :ref:`store <syntax-store>` and a computation over a sequence of remaining :ref:`instructions <syntax-instr>`, :ref:`annotated <notation-attime>` with the :ref:`time <syntax-time>` it was last active.
+
+A *local configuration* describes the state of an active function.
+It consists of the local :ref:`store <syntax-store>` of the respective thread, the :ref:`frame <syntax-frame>` of the function, and the sequence of remaining :ref:`instructions <syntax-instr>` in that function.
+
+.. math::
+   \begin{array}{llcl}
+   \production{(global configuration)} & \config &::=&
+     \sharedstore; \thread^\ast \\
+   \production{(thread)} & \thread &::=&
+     \store; \instr^\ast~\AT~\time \\
+   \production{(local configuration)} & \lconfig &::=&
+     \store; \frame; \instr^\ast \\
+   \end{array}
 
-   It is an invariant of the semantics that the length of the byte vector, divided by page size, never exceeds the maximum size, if present.
+A thread has *terminated* when its instruction sequence has been reduced to a :ref:`result <syntax-result>`,
+that is, either a sequence of :ref:`values <syntax-val>` or to a |TRAP|.
 
 
 Convention
 ..........
 
-* The actions :math:`\ARD_{\ord}` and :math:`\AWR_{\ord}` are abbreviated to :math:just `\ARD` and :math:`\AWR` when :math:`\ord` is :math:`\UNORD`.
+* The meta variable :math:`P` ranges over threads where clear from context.
+
 
+.. index:: ! reduction, configuration, ! termination
 
-.. todo:: define notational shorthands over actions
+Reduction
+~~~~~~~~~
 
+Formally, WebAssembly computation is defined by two *small-step reduction* relations on global and local :ref:`configurations <syntax-config>`
+that define how a single step of execution modifies these configurations, respectively.
 
-.. index:: ! event, action, time stamp
-.. _syntax-evt:
 
-Events
-~~~~~~
+Global Reduction
+................
 
-The (possibly empty) set of :ref:`actions <syntax-act>` performed by an individual :ref:`instruction <syntax-instr>` collectively forms an atomic *event*. Each event is also associated with a :ref:`time stamp <syntax-time>` that uniquely identifies the event.
+*Global reduction* is concerned with synchronization between multiple :ref:`threads <syntax-thread>`.
+It emits a (possibly empty) set of events that are produced by the corresponding step of computation.
+
+Formally, glboal reduction is a relation
 
 .. math::
-   \begin{array}{llcl}
-   \production{(event)} & \evt &::=&
-     \act^\ast~\AT~\time \\
+   \config \stepto^{\evt^\ast} \config
+
+defined by inductive rewrite rules on global configurations.
+
+The following structural rule for global reduction delegates to local reduction for single thread execution:
+
+.. math::
+   \begin{array}{@{}c@{}}
+   \XX{SS}; P_1^\ast~(S; \instr^\ast \AT h)~P_2^\ast
```

Comment:
I might be misreading this rule (I'm having trouble getting the markup building to work) but why is the empty frame `F_\epsilon` needed? I would have thought that each thread would have its own frame (as defined [here](https://webassembly.github.io/spec/core/exec/runtime.html?highlight=thread#configurations)) instead of just being local store + instructions.

Reply:
Well, then you would just permanently keep all the empty frames in the global configuration (and formulate well-formedness rules for them), and you'd need to create them during thread creation.

The main reason I did it like that in the original formulation was that there was only one level of reduction relation, so no natural place to inject the empty frame as you go. But now that we have that, it seems less attractive to keep that artefact around. The only appeal would be that thread and local configuration get unified, but is that valuable? I don't know, though I could be convinced otherwise.

Btw, I don't know if you noticed, but despite the split store, the rule here is currently simpler than the one in our paper:

* It does not allow any allocation, i.e., the global store remains unchanged. At least for now, the only operation that allocates is instantiate, which is simpler to model with a global reduction rule.

* It also doesn't pass the prior global store to the local reduction, since the operations where it would be needed (mutation of a shared table or global) are not currently supported.

So this completely decouples local reduction from the global store for now, which is nice. Even if we add allocation forms, it might actually be simpler to do so through directly through global reductions. (Typed updates will remain an annoyance, though.)

Reply:
My thinking was that unifying threads and local configurations would mean they wouldn't need separate validity rules in the soundness statement. It's a very small thing either way.

Rest of your comment looks good!